credoai/modules/model_modules/privacy.py

```diff
@@ -131,54 +181,73 @@ def _general_attack_method(self, attack_details):
-                x_test_assess,
-                y_train_attack,
-                y_train_assess,
-                y_test_attack,
-                y_test_assess,
-            ) = train_test_split(
-                self.x_train,
-                self.x_test,
-                self.y_train,
-                self.y_test,
-                train_size=self.attack_train_ratio,
-                random_state=42,
+            attack_train_size = int(len(self.x_train) * self.attack_train_ratio)
+            attack_test_size = int(len(self.x_test) * self.attack_train_ratio)
+            (x_train_attack, y_train_attack, x_test_attack, y_test_attack) = (
+                self.x_train[:attack_train_size],
+                self.y_train[:attack_train_size],
+                self.x_test[:attack_test_size],
+                self.y_test[:attack_test_size],
+            )
+            (x_train_assess, y_train_assess, x_test_assess, y_test_assess) = (
+                self.x_train[attack_train_size:],
+                self.y_train[attack_train_size:],
+                self.x_test[attack_test_size:],
+                self.y_test[attack_test_size:],
             )
 
+        # Fit of attack model
+        if attack_details["fit"] == "train_test":
             # Split train and test further and fit the model
             attack.fit(x_train_attack, y_train_attack, x_test_attack, y_test_attack)
 
+        if attack_details["fit"] == "train_only":
+            attack.fit(x_train_assess)
+
+        ## Assessment
         # Sets balancing -> This might become optional if we use other metrics, tbd
         x_train_bln, y_train_bln, x_test_bln, y_test_bln = self._balance_sets(
             x_train_assess, y_train_assess, x_test_assess, y_test_assess
         )
 
         # Attack inference
-        train = attack.infer(x_train_bln, y_train_bln)
-        test = attack.infer(x_test_bln, y_test_bln)
+        if attack_details["assess"] == "membership":
+            train = attack.infer(x_train_bln, y_train_bln)
+            test = attack.infer(x_test_bln, y_test_bln)
+            return self._assess_attack(train, test, accuracy_score)
+
+        if attack_details["assess"] == "attribute":
+            # Compare infered feature with original
+            extra_arg = {}
+            if "estimator" in attack_details["attack"]["kwargs"].keys():
+                extra_arg = {
+                    "pred": np.array(
+                        [
+                            np.argmax(arr)
+                            for arr in self.attack_model.predict(x_test_bln)
+                        ]
+                    ).reshape(-1, 1)
+                }
+            original = x_test_bln[:, self.attack_feature].copy()
+            inferred = attack.infer(
+                np.delete(x_test_bln, self.attack_feature, 1), **extra_arg
+            )
 
-        return self._assess_attack(train, test, accuracy_score)
+            return np.sum(inferred == original) / len(inferred)
```

Comment:
What will happen if there are multiple features? I think this calculation is wrong - this should be divided by the number of elements, not just the length. E.g., `np.size(inferred)`

However, even that just reports the attribute inference across all items. We may want to report attribute inference success differentially for each value. This form of the test says "if I have none of these attributes, how well can the system infer them". Do you think that's the better test than "if I have all but one of these features, how well can I infer that feature"?